p2p/discovery/message.go

```diff
@@ -18,39 +20,189 @@ const (
 	NeighborsMsg MsgType = 5
 )
 
-type Ping struct {
+const (
+	VERSION uint = 1
+)
+
+type ping struct {
 	Version uint // TODO add version check
-	ID      NodeID
+	SelfID  NodeID
+
+	to NodeID
 }
 
-type Pong struct {
-	ID NodeID
+type pong struct {
+	SelfID NodeID
 }
 
-type FindNode struct {
-	Target NodeID
+type findNode struct {
+	SelfID  NodeID
+	QueryID NodeID
+
+	to NodeID
 }
 
-type Neighbors struct {
-	Nodes []RPCNode
+type neighbors struct {
+	SelfID NodeID
+	Nodes  []*rpcNode
 }
 
-type RPCNode struct {
-	ID      NodeID
+type rpcNode struct {
+	SelfID  NodeID
 	IP      net.IP
 	UDPPort uint16
 }
 
-func ByteToMsgType(byte byte) MsgType {
+func (r *rpcNode) ToNode() *Node {
+	return NewNode(r.SelfID, r.IP, r.UDPPort)
+}
+
+func byteToMsgType(byte byte) MsgType {
 	return MsgType(uint8(byte))
 }
 
-func MsgTypeToByte(t MsgType) byte {
+func msgTypeToByte(t MsgType) byte {
 	return byte(t)
 }
 
 func generateBuff(code MsgType, encoding []byte) []byte {
-	buff := []byte{MsgTypeToByte(code)}
+	buff := []byte{msgTypeToByte(code)}
 
 	return append(buff, encoding...)
 }
+
+// handle send ping msg and add pending
+func (m *ping) handle(t *udp, from *net.UDPAddr) {
+	log.Debug("received ping from: %s", common.BytesToHex(m.SelfID.Bytes()))
+
+	// response with pong
+	if m.Version != VERSION {
+		return
+	}
+
+	resp := &pong {
+		SelfID: t.self.ID,
+	}
+
+	t.sendMsg(PongMsg, resp, from)
+}
+
+func (m *ping) send(t *udp, to *net.UDPAddr) {
+	log.Debug("send ping msg to: %s", common.BytesToHex(m.to.Bytes()))
+
+	p := &pending {
+		from: m.to,
+		code: PongMsg,
+
+		callback: func(resp interface{}, addr *net.UDPAddr) (done bool) {
+			r := resp.(*pong)
+			n := NewNodeWithAddr(r.SelfID, addr)
+			t.table.updateNode(n)
+
+			log.Debug("received pong msg: %s", common.BytesToHex(r.SelfID.Bytes()))
+
+			return true
+		},
+		errorCallBack: func() { // delete this node when ping timeout, TODO add time limit
```

Comment:
so when exceptions thrown in callback, then will trap into errorCallBack?

Reply:
errorCallBack calls when response timeout and send failed